### PR TITLE
Add routine for verifying two marker trees are disjoint

### PR DESCRIPTION
**Pull request by ibraheemdev**
## Summary

Implements https://github.com/astral-sh/uv/issues/3355.

### CODE REVIEW

File: crates/uv-resolver/src/marker.rs [hunk 1]
```diff
@@ -0,0 +1,216 @@
...
+use pep440_rs::{Operator, Version, VersionSpecifier};
+use pep508_rs::{
+    ExtraName, ExtraOperator, MarkerExpression, MarkerOperator, MarkerTree, MarkerValueString,
+    MarkerValueVersion,
+};
+
+use crate::pubgrub::PubGrubSpecifier;
+
+/// Returns `true` if there is no environment in which both marker trees can both apply, i.e.
+/// the expression `this and other` is always false.
```
Comment (BurntSushi) [inline crates/uv-resolver/src/marker.rs:13]: Perfect.

**Comment by ibraheemdev** [comment]
This should be ready, but I have a few concerns.

According to PEP508, `python_version` is defined to be to `'.'.join(platform.python_version_tuple()[:2])` (i.e. no patch version, as opposed to `python_full_version`). However, we don't handle this in the existing marker code, so `python_version == '3.7.1'` will fail if the environment contains version `3.7'`. To make this work with pubgrub I believe we have to normailze all marker python versions to `x.y.*`.

The check currently doesn't consider the special case of `python_version` and `python_full_version`, which are related and can be disjoint.

However, I think this can be merged as is and `python_version` markers be fixed separately.

**Review by BurntSushi — APPROVED** [review]
Sweet. The tests look great. This is awesome.

One thing I noticed (but didn't realize yesterday) is that this is in `uv-resolver`. I think that's fine because it should be the only place we need it. But it does seem like it would be nice for this to live in the `pep508_rs` crate. We'd have to change the implementation to not use `PubGrubSpecifier` though. But, I think it's fine for it to live here until there's a use case to move it. :-)

**Comment by BurntSushi** [comment]
> However, I think this can be merged as is and `python_version` markers be fixed separately.

Yeah I think this is fine.

**Comment by ibraheemdev** [comment]
I originally did have this in `pep508_rs`, but moved it because of how much simpler it would be to use `pubgrub`.

**Pull request merged**